src/crypto/pubkeyEncryptor.ts

```diff
@@ -0,0 +1,43 @@
+import nacl from "tweetnacl";
+import ed2curve from "ed2curve";
+import crypto, {sign} from "crypto";
+import {X25519EncryptedData} from "./x25519EncryptedData";
+import {UserPublicKey, UserSecretKey} from "../userKeys";
+import {PubKeyEncCipher, PubKeyEncNonceLength, PubKeyEncVersion} from "./constants";
+
+export class PubkeyEncryptor {
+    static encrypt(data: Buffer, recipientPubKey: UserPublicKey, authSecretKey: UserSecretKey): X25519EncryptedData {
+        // create a new x225519 keypair that will be used for EDH
+        const edhPair = nacl.sign.keyPair();
+        const recipientDHPubKey = ed2curve.convertPublicKey(recipientPubKey.valueOf());
+        if (recipientDHPubKey === null) {
+            throw new Error("Could not convert ed25519 public key to x25519");
+        }
+        const edhConvertedSecretKey = ed2curve.convertSecretKey(edhPair.secretKey);
+
+        const nonce = nacl.randomBytes(PubKeyEncNonceLength);
+        const encryptedMessage = nacl.box(data, nonce, recipientDHPubKey, edhConvertedSecretKey);
+
+        // Note that the ciphertext is already authenticated for the ephemeral key - but we want it authenticated by
+        //  the an elrond ed25519 key which the user interacts with. A signature over H(ciphertext | edhPubKey)
```

Comment:
Typo in comment.

Reply:
`edhPair` is, therefore, an ed25519 key-pair generated on-the-fly (ephemeral). Public key of Bob (not ephemeral) is used to encrypt the `data`; secret part of ephemeral key is used to sign the encrypted data. Secret key of Alice is used to apply a "wrapping" signature. The public part of the ephemeral key is saved in the JSON object (non-repudiable especially due to the "wrapping" signature above).

Is the summary correct?

Not sure, wasn't direct encryption (`nacl.box`) with Bob's public key a viable alternative (no ephemeral key etc.)?

Reply:
Not only signing, it's also used for Diffie Hellman. That's one of the reasons for the ephemeral key. Will try to describe below how the encryption works.

First, even though it's public key encryption, you end up using symetric key that can be computed by both parties (Bob - the receiver and Alice the sender). I will omit the ed25519/x25519 conversions

Bob has keypair `(b, B)` where:
```
B = b * G
```
Alice computes ephemeral `(e, E)` where:
```
E = e * G
```

The secret used for encryption is `x = b * e * G`:
```
// Computed by Alice
x = e * B
// Computed by Bob
x = b * E
```

So the secret part of the ephemeral key is used for authentication but most important, it's also part of the secret used to encrypt the data.

Without the ephemeral key, you will have to encrypt using Alice's private key which would have security implications (dual-purpose use of a private key, also you will have to keep track of the nonce used for encryption)